text/0000-read-all.md

```diff
@@ -0,0 +1,50 @@
+- Feature Name: read_all
+- Start Date: 2015-03-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Rust's Write trait has write_all, which attempts to write an entire
+buffer.  This proposal adds read_all, which attempts to read a fixed
+number of bytes into a given buffer.
+
+# Motivation
+
+The new read_all method will allow programs to read from disk without
+having to write their own read loops.  Most Rust programs which need
+to read from disk will prefer this to the plain read function.  Many C
+programs have the same need, and solve it the same way (e.g. git has
+read_in_full).  Here's one example of a Rust library doing this:
+https://github.com/BurntSushi/byteorder/blob/master/src/new.rs#L184
+
+# Detailed design
+
+The read_all function will take a mutable, borrowed slice of u8 to
+read into, and will attempt to fill that entire slice with data.
+
+It will loop, calling read() once per iteration and attempting to read
+the remaining amount of data.  If read returns EINTR, the loop will
+retry.  If there are no more bytes to read (as signalled by a return
+of Ok(0) from read()), a new error type, ErrorKind::ReadZero, will be
+returned.  In the event of another error, that error will be
+returned. After a read call returns having successfully read some
+bytes, the total number of bytes read will be updated.  If that
+total is equal to the size of the buffer, read will return
+successfully.
```

Comment:
I think `ShortRead` (or maybe `UnexpectedEof`) is a much better name for the error than `ReadZero`. I agree that the success case should be limited to completely filling the buffer, but I also agree with @ArtemGr that the error in the case of a premature EOF should include the number of bytes successfully read. I don't think there is anything analogous to running out of data in `write_all` to compare to: when writing, either you successfully write everything, or some IO error occurs.

I agree that for other errors (actual IO errors), there's probably no need to return the number of bytes read. As you note, this is analogous to `write_all`, and the user can always call `read` directly if they need to know that information.

Otherwise, :+1: to having a read_all method. (I have similarly needed to implemented this functionality for my current project, so it seems quite a useful thing to have in the standard library.)